cmd/snap-recovery/volmgr/sfdisk.go

```diff
@@ -0,0 +1,260 @@
+// in usedPartitions. Return a role to device node map.
+func (sf *SFDisk) CreatePartitions(pv *gadget.LaidOutVolume, usedPartitions []bool) (map[string]string, error) {
+	buf, deviceMap := buildPartitionList(sf.partitionTable, pv, usedPartitions)
+
+	// Write the partition table
+	cmd := exec.Command("sfdisk", sf.device)
+	cmd.Stdin = buf
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return deviceMap, osutil.OutputErr(output, err)
+	}
+
+	// Reload the partition table
+	if output, err := exec.Command("partx", "-u", sf.device).CombinedOutput(); err != nil {
+		return deviceMap, osutil.OutputErr(output, fmt.Errorf("cannot update partition table: %s", err))
+	}
+
+	return deviceMap, nil
+}
+
+// positionedVolumeFromDump takes an sfdisk dump format and returns the partitioning
+// information as a laid out volume.
+func positionedVolumeFromDump(dump *SFDiskDeviceDump) (*gadget.LaidOutVolume, error) {
+	ptable := dump.PartitionTable
+
+	if ptable.Unit != "sectors" {
+		return nil, fmt.Errorf("cannot position partitions: unknown unit %q", ptable.Unit)
+	}
+
+	structure := make([]gadget.VolumeStructure, len(ptable.Partitions))
+	ps := make([]gadget.LaidOutStructure, len(ptable.Partitions))
+
+	for i, p := range ptable.Partitions {
+		info, err := filesystemInfo(p.Node)
+		if err != nil {
+			return nil, fmt.Errorf("cannot obtain filesystem information: %v", err)
+		}
+		if len(info.BlockDevices) < 1 {
+			continue
+		}
+		bd := info.BlockDevices[0]
+
+		structure[i] = gadget.VolumeStructure{
+			Name:       p.Name,
+			Size:       gadget.Size(p.Size) * sectorSize,
+			Label:      bd.Label,
+			Type:       p.Type,
+			Filesystem: bd.FSType,
+		}
+		ps[i] = gadget.LaidOutStructure{
+			VolumeStructure: &structure[i],
+			StartOffset:     gadget.Size(p.Start) * sectorSize,
+			Index:           i + 1,
+		}
+	}
+
+	pv := &gadget.LaidOutVolume{
+		Volume: &gadget.Volume{
+			ID:        ptable.ID,
+			Structure: structure,
+		},
+		Size:             gadget.Size(ptable.LastLBA),
+		SectorSize:       sectorSize,
+		LaidOutStructure: ps,
+	}
+
+	return pv, nil
+}
+
+func deviceName(name string, index int) string {
+	if len(name) > 0 {
+		last := name[len(name)-1]
+		if last >= '0' && last <= '9' {
+			return fmt.Sprintf("%sp%d", name, index)
+		}
+	}
+	return fmt.Sprintf("%s%d", name, index)
+}
+
+// buildPartitionList builds a list of partitions based on the current device contents
+// and gadget structure list, in sfdisk dump format. Return a partitioning description
+// suitable for sfdisk input and a role to device node map.
+func buildPartitionList(ptable *SFDiskPartitionTable, pv *gadget.LaidOutVolume, usedPartitions []bool) (*bytes.Buffer, map[string]string) {
+	buf := &bytes.Buffer{}
+	deviceMap := map[string]string{}
+
+	// Write partition data in sfdisk dump format
+	fmt.Fprintf(buf, "label: %s\nlabel-id: %s\ndevice: %s\nunit: %s\nfirst-lba: %d\nlast-lba: %d\n\n",
```

Comment:
As noted earlier in the comment about --append, I hope that sfdisk is smart enough to not try to modify this on the existing block device if the values aren't changed.